SRA_ACCESSIONS = [l.strip() for l in open("../SRR_Acc_List.txt").readlines()]
OUT_DIR = "/data/scratch/sturm/pipeline_test/snakemake"

rule fastq:
  """run download"""
  input:
    expand(OUT_DIR + "/fastq/{accession}_pass_1.fastq.gz", accession=SRA_ACCESSIONS)

rule rsem:
  """run rsem"""
  input:
    expand(OUT_DIR + "/rsem/{accession}.genes.results", accession=SRA_ACCESSIONS)

rule download_fastq:
  """download fastq"""
  output:
    OUT_DIR + "/fastq/{accession}_pass_1.fastq.gz"
  params:
    accession="{accession}"
  conda:
    "envs/fastq.yml"
  threads: 1
  shell:
    "fastq-dump --outdir " + OUT_DIR + "/fastq/ "
    "--gzip --skip-technical --readids "
    "--read-filter pass --dumpbase --split-files --clip {params.accession}"

rule run_rsem:
  """align using rsem"""
  threads: 8
  input:
    OUT_DIR + "/fastq/{accession}_pass_1.fastq.gz"
  output:
    OUT_DIR + "/rsem/{accession}.genes.results",
    OUT_DIR + "/rsem/{accession}.isoforms.results"
  params:
    accession="{accession}"
  conda:
    "envs/mapping.yml"
  shell:
    "rsem-calculate-expression -p 8 "
    " --star --star-gzipped-read-file "
    " --no-bam-output --append-names "
    " --single-cell-prior "
    " {input} "
    " /home/sturm/scratch/pipeline_test/rsem/ref_hg38_ensemble "
    " " + OUT_DIR + "/rsem/{params.accession} "
